src/main/java/com/microsoft/sqlserver/jdbc/SQLServerTimeoutManager.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Microsoft JDBC Driver for SQL Server Copyright(c) Microsoft Corporation All rights reserved. This program is made
+ * available under the terms of the MIT License. See the LICENSE file in the project root for more information.
+ */
+
+package com.microsoft.sqlserver.jdbc;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+/**
+ * Object that handles management of timeout tasks across all sql server connections
+ */
+final class SQLServerTimeoutManager {
+    private static ScheduledExecutorService scheduledTimeoutTasks = Executors.newScheduledThreadPool(1,
+            createThreadFactory("com.microsoft.sqlserver.jdbc.SQLServerTimeoutManager"));
+    private static ExecutorService timeoutTaskWorker = Executors.newSingleThreadExecutor(
+            createThreadFactory("com.microsoft.sqlserver.jdbc.SQLServerTimeoutManager.TimeoutTaskWorker"));
+    private static Map<UUID, List<TimeoutCommand<?>>> timeoutCommands = new HashMap<>();
+    private final static Logger logger = Logger.getLogger("com.microsoft.sqlserver.jdbc.SQLServerTimeoutManager");
+
+    static void startTimeoutCommand(TimeoutCommand<?> timeoutCommand) {
+        if (scheduledTimeoutTasks.isShutdown()) {
+            // reset id counter for timeout commands
+            TimeoutCommand.uniqueId.set(0);
+            scheduledTimeoutTasks = Executors.newScheduledThreadPool(1);
+            timeoutTaskWorker = Executors.newSingleThreadExecutor();
+        }
+
+        timeoutCommand.setTimeoutTask(scheduledTimeoutTasks.schedule(new Runnable() {
+
+            @Override
+            public void run() {
+                Future<?> timeoutTask = timeoutTaskWorker.submit(new Runnable() {
+                    @Override
+                    public void run() {
+                        timeoutCommand.interrupt();
+                    }
+                });
+                try {
+                    // if the timeout command takes too long to interrupt, cancel it and release
+                    timeoutTask.get(10, TimeUnit.SECONDS);

```

Comment:
This shouldn't be an arbitrary number, will think about what should be done here.